File: app/src/main/java/com/example/smartswine/utils/SwineGrowthDatabase.kt
package com.example.smartswine.utils

import java.util.Locale

data class GrowthPoint(
    val ageDays: Int,
    val weightKg: Double
)

data class BreedCurve(
    val key: String,
    val displayName: String,
    val points: List<GrowthPoint>
)

object SwineGrowthDatabase {

    // 10% reduced curves from veterinary standards
    val LargeWhite = BreedCurve(
        key = "Large White",
        displayName = "Large White",
        points = listOf(
            GrowthPoint(0, 1.26),
            GrowthPoint(28, 7.20),
            GrowthPoint(56, 19.80),
            GrowthPoint(84, 39.15),
            GrowthPoint(112, 63.90),
            GrowthPoint(140, 87.75),
            GrowthPoint(168, 105.75),
            GrowthPoint(252, 144.00),
            GrowthPoint(364, 180.00)
        )
    )

    val Landrace = BreedCurve(
        key = "Landrace",
        displayName = "Landrace",
        points = listOf(
            GrowthPoint(0, 1.22),
            GrowthPoint(28, 6.93),
            GrowthPoint(56, 19.35),
            GrowthPoint(84, 37.80),
            GrowthPoint(112, 62.10),
            GrowthPoint(140, 85.50),
            GrowthPoint(168, 103.50),
            GrowthPoint(252, 139.50),
            GrowthPoint(364, 171.00)
        )
    )

    val Duroc = BreedCurve(
        key = "Duroc",
        displayName = "Duroc",
        points = listOf(
            GrowthPoint(0, 1.31),
            GrowthPoint(28, 7.47),
            GrowthPoint(56, 21.60),
            GrowthPoint(84, 41.40),
            GrowthPoint(112, 67.50),
            GrowthPoint(140, 91.80),
            GrowthPoint(168, 112.50),
            GrowthPoint(252, 153.00),
            GrowthPoint(364, 189.00)
        )
    )

    val Hampshire = BreedCurve(
        key = "Hampshire",
        displayName = "Hampshire",
        points = listOf(
            GrowthPoint(0, 1.17),
            GrowthPoint(28, 6.75),
            GrowthPoint(56, 18.00),
            GrowthPoint(84, 36.00),
            GrowthPoint(112, 58.50),
            GrowthPoint(140, 81.00),
            GrowthPoint(168, 99.00),
            GrowthPoint(252, 130.50),
            GrowthPoint(364, 162.00)
        )
    )

    val Berkshire = BreedCurve(
        key = "Berkshire",
        displayName = "Berkshire",
        points = listOf(
            GrowthPoint(0, 1.13),
            GrowthPoint(28, 6.30),
            GrowthPoint(56, 17.10),
            GrowthPoint(84, 34.20),
            GrowthPoint(112, 54.00),
            GrowthPoint(140, 76.50),
            GrowthPoint(168, 94.50),
            GrowthPoint(252, 126.00),
            GrowthPoint(364, 153.00)
        )
    )

    val LocalHeritage = BreedCurve(
        key = "Local / Heritage",
        displayName = "Local / Heritage",
        points = listOf(
            GrowthPoint(0, 0.81),
            GrowthPoint(28, 4.50),
            GrowthPoint(56, 9.90),
            GrowthPoint(84, 17.10),
            GrowthPoint(112, 27.00),
            GrowthPoint(140, 36.90),
            GrowthPoint(168, 46.80),
            GrowthPoint(252, 67.50),
            GrowthPoint(364, 90.00)
        )
    )


    val standardOptions = listOf(
        "Large White",
        "Landrace",
        "Duroc",
        "Hampshire",
        "Berkshire",
        "Large White x Landrace (F1)",
        "Duroc x (Large White x Landrace)",
        "Duroc x Landrace",
        "Duroc x Large White",
        "Hampshire x Landrace",
        "Hampshire x Large White",
        "Local / Heritage",
        "Other"
    )

    /**
     * Resolves a breed name string to a BreedCurve.
     * Supports dynamic cross-breed blending if parent breeds are separated by 'x', '/' or '+'.
     */
    fun resolveBreedCurve(breedInput: String): BreedCurve {
        val normalized = breedInput.trim().lowercase(Locale.ROOT)

        // 0. Handle empty or unknown breed - Use Local / Heritage as the least performing standard
        if (normalized.isEmpty() || normalized == "other" || normalized == "none" || normalized == "unknown") {
            return LocalHeritage
        }

        // 1. Exact or keyword checks for predefined crosses
        if (normalized.contains("duroc") && normalized.contains("large white") && normalized.contains("landrace")) {
            return blendCurves(listOf(Duroc, LargeWhite, Landrace), "Duroc x (Large White x Landrace)", 1.08) // 8% heterosis
        }
        if (normalized.contains("large white") && normalized.contains("landrace")) {
            return blendCurves(listOf(LargeWhite, Landrace), "Large White x Landrace (F1)", 1.05) // 5% heterosis
        }

        // 2. Parse cross-breeding tokens
        val separators = regexSeparators()
        val parts = normalized.split(separators).map { it.trim() }.filter { it.isNotEmpty() }

        if (parts.size > 1) {
            val matchedCurves = parts.mapNotNull { findBaseBreed(it) }
            if (matchedCurves.isNotEmpty()) {
                val crossName = matchedCurves.joinToString(" x ") { it.displayName }
                return blendCurves(matchedCurves, crossName, 1.05) // 5% heterosis
            }
        }

        // 3. Match single base breed keyword
        val singleMatch = findBaseBreed(normalized)
        if (singleMatch != null) {
            return singleMatch
        }

        // 4. Fallback to Local / Heritage standard if no breed is selected or matched
        return LocalHeritage
    }

    private fun findBaseBreed(token: String): BreedCurve? {
        return when {
            token.contains("duroc") -> Duroc
            token.contains("landrace") -> Landrace
            token.contains("large white") || token.contains("yorkshire") || token.contains("white") -> LargeWhite
            token.contains("hampshire") || token.contains("hamp") -> Hampshire
            token.contains("berkshire") || token.contains("berk") -> Berkshire
            token.contains("local") || token.contains("heritage") || token.contains("indigenous") || token.contains("native") || token.contains("kolbroek") || token.contains("mukota") -> LocalHeritage
            else -> null
        }
    }

    private fun regexSeparators() = Regex("[x/+]|\\bcross\\b|\\bhybrid\\b")

    /**
     * Blends multiple growth curves by averaging the expected weights at each standard age milestone.
     */
    private fun blendCurves(curves: List<BreedCurve>, displayName: String, vigorFactor: Double = 1.0): BreedCurve {
        val standardAges = listOf(0, 28, 56, 84, 112, 140, 168, 252, 364)
        val blendedPoints = standardAges.map { age ->
            val sumWeights = curves.sumOf { curve ->
                interpolateCurve(curve.points, age)
            }
            val avgWeight = sumWeights / curves.size
            GrowthPoint(age, (avgWeight * vigorFactor).round(2))
        }
        return BreedCurve(
            key = displayName,
            displayName = displayName,
            points = blendedPoints
        )
    }

    /**
     * Estimates weight for a given age using linear interpolation between known data points.
     */
    fun interpolateCurve(points: List<GrowthPoint>, ageDays: Int): Double {
        if (points.isEmpty()) return 0.0
        val sorted = points.sortedBy { it.ageDays }
        
        // Exact age matches or edge limits
        if (ageDays <= sorted.first().ageDays) return sorted.first().weightKg
        if (ageDays >= sorted.last().ageDays) return sorted.last().weightKg

        // Find intervals to interpolate
        for (i in 0 until sorted.size - 1) {
            val p1 = sorted[i]
            val p2 = sorted[i + 1]
            if (ageDays in p1.ageDays..p2.ageDays) {
                val ratio = (ageDays - p1.ageDays).toDouble() / (p2.ageDays - p1.ageDays).toDouble()
                return p1.weightKg + ratio * (p2.weightKg - p1.weightKg)
            }
        }
        return sorted.last().weightKg
    }

    /**
     * Scores the pig's performance based on age (days) and weight.
     * Returns "Excellent", "Good", "Caution", "Poor" or "Blank" (if weight is 0 or age is invalid).
     */
    fun evaluatePerformance(breed: String, ageDays: Int, actualWeight: Double): String {
        if (actualWeight <= 0.0) {
            return "Blank"
        }
        val safeAgeDays = if (ageDays < 0) 0 else ageDays
        val curve = resolveBreedCurve(breed)
        val expectedWeight = interpolateCurve(curve.points, safeAgeDays)
        val safeExpectedWeight = if (expectedWeight <= 0.0) {
            curve.points.firstOrNull()?.weightKg ?: 1.0
        } else {
            expectedWeight
        }

        val ratio = actualWeight / safeExpectedWeight
        
        // For mature pigs (age > 270 days / 9 months), growth plateaus.
        // We broaden the "Good" range to account for normal variations in mature size
        // without falsely flagging healthy mature animals whose growth has slowed down.
        return if (safeAgeDays > 270) {
            when {
                ratio >= 1.35 -> "Excellent"
                ratio >= 0.85 -> "Good"
                ratio >= 0.70 -> "Caution"
                else -> "Poor"
            }
        } else {
            when {
                ratio >= 1.15 -> "Excellent"
                ratio >= 0.90 -> "Good"
                ratio >= 0.75 -> "Caution"
                else -> "Poor"
            }
        }
    }

    private fun Double.round(decimals: Int): Double {
        var multiplier = 1.0
        repeat(decimals) { multiplier *= 10.0 }
        return kotlin.math.round(this * multiplier) / multiplier
    }
}

File: web/src/lib/swineGrowthDatabase.ts
interface GrowthPoint {
  ageDays: number;
  weightKg: number;
}

interface BreedCurve {
  key: string;
  displayName: string;
  points: GrowthPoint[];
}

const LargeWhite: BreedCurve = {
  key: "Large White",
  displayName: "Large White",
  points: [
    { ageDays: 0, weightKg: 1.26 },
    { ageDays: 28, weightKg: 7.20 },
    { ageDays: 56, weightKg: 19.80 },
    { ageDays: 84, weightKg: 39.15 },
    { ageDays: 112, weightKg: 63.90 },
    { ageDays: 140, weightKg: 87.75 },
    { ageDays: 168, weightKg: 105.75 },
    { ageDays: 252, weightKg: 144.00 },
    { ageDays: 364, weightKg: 180.00 }
  ]
};

const Landrace: BreedCurve = {
  key: "Landrace",
  displayName: "Landrace",
  points: [
    { ageDays: 0, weightKg: 1.22 },
    { ageDays: 28, weightKg: 6.93 },
    { ageDays: 56, weightKg: 19.35 },
    { ageDays: 84, weightKg: 37.80 },
    { ageDays: 112, weightKg: 62.10 },
    { ageDays: 140, weightKg: 85.50 },
    { ageDays: 168, weightKg: 103.50 },
    { ageDays: 252, weightKg: 139.50 },
    { ageDays: 364, weightKg: 171.00 }
  ]
};

const Duroc: BreedCurve = {
  key: "Duroc",
  displayName: "Duroc",
  points: [
    { ageDays: 0, weightKg: 1.31 },
    { ageDays: 28, weightKg: 7.47 },
    { ageDays: 56, weightKg: 21.60 },
    { ageDays: 84, weightKg: 41.40 },
    { ageDays: 112, weightKg: 67.50 },
    { ageDays: 140, weightKg: 91.80 },
    { ageDays: 168, weightKg: 112.50 },
    { ageDays: 252, weightKg: 153.00 },
    { ageDays: 364, weightKg: 189.00 }
  ]
};

const Hampshire: BreedCurve = {
  key: "Hampshire",
  displayName: "Hampshire",
  points: [
    { ageDays: 0, weightKg: 1.17 },
    { ageDays: 28, weightKg: 6.75 },
    { ageDays: 56, weightKg: 18.00 },
    { ageDays: 84, weightKg: 36.00 },
    { ageDays: 112, weightKg: 58.50 },
    { ageDays: 140, weightKg: 81.00 },
    { ageDays: 168, weightKg: 99.00 },
    { ageDays: 252, weightKg: 130.50 },
    { ageDays: 364, weightKg: 162.00 }
  ]
};

const Berkshire: BreedCurve = {
  key: "Berkshire",
  displayName: "Berkshire",
  points: [
    { ageDays: 0, weightKg: 1.13 },
    { ageDays: 28, weightKg: 6.30 },
    { ageDays: 56, weightKg: 17.10 },
    { ageDays: 84, weightKg: 34.20 },
    { ageDays: 112, weightKg: 54.00 },
    { ageDays: 140, weightKg: 76.50 },
    { ageDays: 168, weightKg: 94.50 },
    { ageDays: 252, weightKg: 126.00 },
    { ageDays: 364, weightKg: 153.00 }
  ]
};

const LocalHeritage: BreedCurve = {
  key: "Local / Heritage",
  displayName: "Local / Heritage",
  points: [
    { ageDays: 0, weightKg: 0.81 },
    { ageDays: 28, weightKg: 4.50 },
    { ageDays: 56, weightKg: 9.90 },
    { ageDays: 84, weightKg: 17.10 },
    { ageDays: 112, weightKg: 27.00 },
    { ageDays: 140, weightKg: 36.90 },
    { ageDays: 168, weightKg: 46.80 },
    { ageDays: 252, weightKg: 67.50 },
    { ageDays: 364, weightKg: 90.00 }
  ]
};

export function interpolateCurve(points: GrowthPoint[], ageDays: number): number {
  if (points.length === 0) return 0;
  const sorted = [...points].sort((a, b) => a.ageDays - b.ageDays);

  if (ageDays <= sorted[0].ageDays) return sorted[0].weightKg;
  if (ageDays >= sorted[sorted.length - 1].ageDays) return sorted[sorted.length - 1].weightKg;

  for (let i = 0; i < sorted.length - 1; i++) {
    const p1 = sorted[i];
    const p2 = sorted[i + 1];
    if (ageDays >= p1.ageDays && ageDays <= p2.ageDays) {
      const ratio = (ageDays - p1.ageDays) / (p2.ageDays - p1.ageDays);
      return p1.weightKg + ratio * (p2.weightKg - p1.weightKg);
    }
  }
  return sorted[sorted.length - 1].weightKg;
}

function blendCurves(curves: BreedCurve[], displayName: string, vigorFactor = 1.0): BreedCurve {
  const standardAges = [0, 28, 56, 84, 112, 140, 168, 252, 364];
  const blendedPoints = standardAges.map((age) => {
    let sumWeights = 0;
    curves.forEach((curve) => {
      sumWeights += interpolateCurve(curve.points, age);
    });
    const avgWeight = sumWeights / curves.length;
    return {
      ageDays: age,
      weightKg: Math.round(avgWeight * vigorFactor * 100) / 100,
    };
  });
  return {
    key: displayName,
    displayName,
    points: blendedPoints,
  };
}

function findBaseBreed(token: string): BreedCurve | null {
  const t = token.toLowerCase();
  if (t.includes("duroc")) return Duroc;
  if (t.includes("landrace")) return Landrace;
  if (t.includes("large white") || t.includes("yorkshire") || t.includes("white")) return LargeWhite;
  if (t.includes("hampshire") || t.includes("hamp")) return Hampshire;
  if (t.includes("berkshire") || t.includes("berk")) return Berkshire;
  if (
    t.includes("local") ||
    t.includes("heritage") ||
    t.includes("indigenous") ||
    t.includes("native") ||
    t.includes("kolbroek") ||
    t.includes("mukota")
  ) {
    return LocalHeritage;
  }
  return null;
}

export function resolveBreedCurve(breedInput: string): BreedCurve {
  const normalized = breedInput.trim().toLowerCase();

  if (!normalized || normalized === "other" || normalized === "none" || normalized === "unknown") {
    return LocalHeritage;
  }

  // Predefined crosses
  if (normalized.includes("duroc") && normalized.includes("large white") && normalized.includes("landrace")) {
    return blendCurves([Duroc, LargeWhite, Landrace], "Duroc x (Large White x Landrace)", 1.08);
  }
  if (normalized.includes("large white") && normalized.includes("landrace")) {
    return blendCurves([LargeWhite, Landrace], "Large White x Landrace (F1)", 1.05);
  }

  // Separators checking
  const separators = /[x/+]|\bcross\b|\bhybrid\b/g;
  const parts = normalized.split(separators).map((s) => s.trim()).filter((s) => s.length > 0);

  if (parts.length > 1) {
    const matchedCurves: BreedCurve[] = [];
    parts.forEach((p) => {
      const b = findBaseBreed(p);
      if (b) matchedCurves.push(b);
    });
    if (matchedCurves.length > 0) {
      const crossName = matchedCurves.map((c) => c.displayName).join(" x ");
      return blendCurves(matchedCurves, crossName, 1.05);
    }
  }

  const singleMatch = findBaseBreed(normalized);
  if (singleMatch) return singleMatch;

  return LocalHeritage;
}

export function evaluatePerformance(breed: string, ageDays: number, actualWeight: number): string {
  if (actualWeight <= 0 || !actualWeight) {
    return "Blank";
  }
  const safeAgeDays = ageDays < 0 ? 0 : ageDays;
  const curve = resolveBreedCurve(breed);
  const expectedWeight = interpolateCurve(curve.points, safeAgeDays);
  const safeExpectedWeight = expectedWeight <= 0
    ? (curve.points[0]?.weightKg || 1.0)
    : expectedWeight;

  const ratio = actualWeight / safeExpectedWeight;

  if (safeAgeDays > 270) {
    if (ratio >= 1.35) return "Excellent";
    if (ratio >= 0.85) return "Good";
    if (ratio >= 0.70) return "Caution";
    return "Poor";
  } else {
    if (ratio >= 1.15) return "Excellent";
    if (ratio >= 0.90) return "Good";
    if (ratio >= 0.75) return "Caution";
    return "Poor";
  }
}

export function parseSwineDate(dateStr?: string): Date | null {
  if (!dateStr) return null;
  const trimmed = dateStr.trim();

  // 1. Check if format is yyyy-mm-dd (e.g. 2026-05-10)
  if (/^\d{4}-\d{2}-\d{2}$/.test(trimmed)) {
    const parts = trimmed.split("-").map(Number);
    const d = new Date(parts[0], parts[1] - 1, parts[2]);
    if (!isNaN(d.getTime())) return d;
  }

  // 2. Check if format is dd/mm/yyyy (e.g. 10/05/2026 or 10-05-2026)
  if (/^\d{2}[/-]\d{2}[/-]\d{4}$/.test(trimmed)) {
    const separator = trimmed.includes("/") ? "/" : "-";
    const parts = trimmed.split(separator).map(Number);
    const d = new Date(parts[2], parts[1] - 1, parts[0]);
    if (!isNaN(d.getTime())) return d;
  }

  // 3. Fallback to standard new Date()
  const d = new Date(trimmed);
  return isNaN(d.getTime()) ? null : d;
}

export function calculateAgeDays(birthDateStr?: string): number {
  if (!birthDateStr) return 0;
  try {
    const birth = parseSwineDate(birthDateStr);
    if (!birth) return 0;
    const today = new Date();
    today.setHours(0, 0, 0, 0);
    const birthMidnight = new Date(birth);
    birthMidnight.setHours(0, 0, 0, 0);
    const diffTime = today.getTime() - birthMidnight.getTime();
    return Math.floor(diffTime / (1000 * 60 * 60 * 24));
  } catch (e) {
    return 0;
  }
}

export function calculateAgeMonths(birthDateStr?: string): number {
  if (!birthDateStr) return 0;
  try {
    const birth = parseSwineDate(birthDateStr);
    if (!birth) return 0;
    const today = new Date();

    let months = (today.getFullYear() - birth.getFullYear()) * 12;
    months += today.getMonth() - birth.getMonth();

    if (today.getDate() < birth.getDate()) {
      months--;
    }
    return Math.max(0, months);
  } catch (e) {
    return 0;
  }
}
